### DISCUSSION TITLE
Bug with multiple screen transitions in UIKit?

### DISCUSSION THREAD

**Discussion by dazy1030** [pointfreeco/swift-composable-architecture#2985] [General]
Hi, I'm just checking the specifications for screen transitions in UINavigationController using the observe method, and I found a bug-like behaviour.
If I create multiple screen transition origins with a single observe method, the first transition I describe will immediately return to the screen from which I made the transition, as shown in the attached video.

https://github.com/pointfreeco/swift-composable-architecture/assets/21330273/94e45408-23a8-4657-9f33-ef691308c0ee

I have prepared a minimal sample that confirms this problem.
https://github.com/dazy1030/UIKitPlayground/tree/tca/multi-navigation-observe

I have confirmed that this problem can be avoided by writing only one transition in the observe method and using two observe methods.

Thanks.

Environments
macOS 14.2.1, Xcode 15.2, iPhone 15 Pro(17.2 simulator)

**Comment by mbrandonw** [comment]
Hi @dazy1030, this is happening because of these lines here:

```swift
if let store = store.scope(state: \.second, action: \.second.presented) {
    let vc = SecondViewController(store: store)
    navigationController?.pushViewController(vc, animated: true)
} else {
    navigationController?.popToViewController(self, animated: true)
}
if let store = store.scope(state: \.third, action: \.third.presented) {
    let vc = ThirdViewController(store: store)
    navigationController?.pushViewController(vc, animated: true)
} else {
    navigationController?.popToViewController(self, animated: true)
}
```

When the `second` state becomes non-`nil` the `SecondViewController` is pushed onto the nav controller, but then immediately after it is popped off because `third == nil`.

You need to follow the pattern we show in our Tic-Tac-Toe demo:

https://github.com/pointfreeco/swift-composable-architecture/blob/fd4b169e212742e201de42c9977bf142dc07ff98/Examples/TicTacToe/tic-tac-toe/Sources/LoginUIKit/LoginViewController.swift#L101-L122

You must keep track of the second and third view controllers so that you know whether or not it is currently presented. Here is how it should look in your situation:

```swift
var secondViewController: SecondViewController?
var thirdViewController: ThirdViewController?

observe { [weak self] in
  guard let self else { return }

  if 
    secondViewController == nil,
    let store = store.scope(state: \.second, action: \.second.presented)
  {
    let vc = SecondViewController(store: store)
    secondViewController = vc
    navigationController?.pushViewController(vc, animated: true)
  } else if secondViewController != nil, store.second == nil {
    navigationController?.popToViewController(self, animated: true)
    secondViewController = nil
  }

  if
    thirdViewController == nil,
    let store = store.scope(state: \.third, action: \.third.presented)
  {
    let vc = ThirdViewController(store: store)
    thirdViewController = vc
    navigationController?.pushViewController(vc, animated: true)
  } else if thirdViewController != nil, store.third == nil {
    navigationController?.popToViewController(self, animated: true)
    thirdViewController = nil
  }
}
```

I will also say that this is not correct:

```swift
if !isMovingToParent {
  store.send(.someViewDismissed)
}
```

This is going to dismiss _both_ controllers whenever this controller appears again. Again I encourage you to look at how we do this in the Tic-Tac-Toe demo:

https://github.com/pointfreeco/swift-composable-architecture/blob/fd4b169e212742e201de42c9977bf142dc07ff98/Examples/TicTacToe/tic-tac-toe/Sources/LoginUIKit/LoginViewController.swift#L129-L131

**Reply by dazy1030** [reply]
Hello, mbrandonw.

Oh, it was a mistake in my code. I mistakenly thought that the processing related to a specific screen could be written closed in an if-else statement. I should have remembered that both expressions are evaluated.

> I will also say that this is not correct:

I had just discovered this PR and corrected it. Thank you very much!
https://github.com/pointfreeco/swift-composable-architecture/pull/2986/files

In the latest commit of the sample, I combined it with Destination Reducer based on the advice I received.
https://github.com/dazy1030/UIKitPlayground/tree/tca/multi-navigation-observe
